### ISSUE TITLE
Support scale groups for activities, entities, and orchestrations

### ISSUE THREAD

**Issue by sebastianburckhardt** [Azure/azure-functions-durable-extension#1691]
**Is your feature request related to a problem? Please describe.**

The control settings `maxConcurrentOrchestratorFunctions`  and `maxConcurrentActivityFunctions` are a bit too basic.
It is not easy to understand how to set these correctly, and it is not always possible.

For example, an application that runs highly CPU-intensive activities has to choose a low value for `maxConcurrentActivityFunctions`; but an application with activities that perform fast asynchronous I/O, such as calls to external services, needs a high value. Similarly, an application that does significant CPU work inside entities should use a low `maxConcurrentOrchestratorFunctions` setting; but applications that do not use entities should generally choose a high setting.

Also, using just two controls does not work if an application requires different scale settings for different functions. There is currently no way to support different scale settings for different functions.
 
**Describe the solution you'd like**

I would like to support scale specifications that are more readable and more flexible. After some thought, I think the following mechanism could work. It has two pieces:

1) Functions can be labeled with a **scale group** label that determines how they should be scaled. In C#, this could be an attribute such as
```csharp
  [FunctionName(nameof(SayHello))]
  [FunctionScaleGroup("CPU")]
  public static string MyActivity([ActivityTrigger] string input) { ... }
```
If no group is specified, it is assumed to be "default" which reverts to the legacy settings. 

1) In the host.json, we add a new section for specifying scale groups:
```
{
  "version": "2.0",
  "extensions": {
    "durableTask": {
      "hubName": "MyDFHub",
      "scalegroups": {
        "CPU": "1 per core",
        "SynchronousIO": "10 per core",
        "AsynchronousIO": "100 per core",
      }
    }
  }
}
```

And for convenience I would have the above specs be defaults that are already defined, i.e.
users can use the groups "CPU", "SynchronousIO", and "AsynchronousIO" without first having to define them.
 
**Describe alternatives you've considered**
At first, I though it would be sufficient to provide concurrency limits for each function individually. However, this is not a good approach as multiple functions can compete for the same resource. Also, I like having predefined groups with understandable names for common activity classes, because it helps to explain the feature.

**Additional thoughts**
A future fancier version of this feature could perhaps
- support global limits, not just per core.
- automatically infer the scaling group by observing the behavior of the function

**Comment by davidmrdavid** [comment]
I really like this idea, thank you @sebastianburckhardt. Indeed, it feels like our current configuration options are neither expressive enough to constraint global behavior nor granular enough to control specific functions; this appears to tackle the latter.

I wonder, if we were to take on this solution, would it mean that we're deprecating the existing configuration knobs of maxConcurrentOrchestratorFunctions and maxConcurrentActivityFunctions? I would say yes, as otherwise the new and the old config knobs might interact poorly and performance tuning would get confusing quickly.

**Comment by davidmrdavid** [comment]
I support this idea, it would give us more granular control during performance tuning. In Python, this could be easily implemented via decorators, or alternatively as part of the `function.json` config. With JS, I think only the `function.json` approach would work.

**Comment by ConnorMcMahon** [comment]
A couple of thoughts.

- "ScaleGroups" seems like the wrong terminology here. This is more about concurrency models than scaling models.
- I have a strong preference against implementing our own concurrency switches. Azure Functions is working on their own improvements to the concurrency model, and I would prefer that we bring these ideas to the core Azure Functions extension model and utilize that as opposed to going our own route.

**Comment by sebastianburckhardt** [comment]
The reason for calling this "scale group" is because the concurrency limits drive the scale-out.

Having this be provided by Azure Functions in general would be great.

Though I think we would still need an ability to access these settings from the DF code, in the sense that it is good for the backend to know these limits before it actually schedules the function for execution. It can do a better job distributing the work.

**Comment by ConnorMcMahon** [comment]
Yeah, the idea behind the new concurrency limits in Azure Functions would be that they would provide some interfaces to interact with when deciding when to grab more work, and it would be up to individual triggers to opt-in to using these new throttling mechanisms.

Ideally we would be able to accomplish this in a backend-agnostic way, but that may not be possible.

**Comment by gorillapower** [comment]
Hi All, would this topic be revived with the new release of this feature in the Webjobs SDK? https://github.com/Azure/azure-functions-host/issues/7299

**Comment by davidmrdavid** [comment]
Hi @gorillapower,

Thanks for reaching out. I believe the two topics are related, but not quite the same. To my understanding, the DC work is to reduce the burden of manual tuning the **existing** performance/concurrency knobs in Azure Functions and its extensions, such as Durable Functions. 

That said, as I understand it, this issue thread proposes creating a new feature: annotations for applying different concurrency settings to different Function (activities/entity/orchestrators) declarations. So while it is true that this work would benefit from DC, I don't think it necessarily unblocks this proposal. That said, I would personally love to have this proposal implemented in the near future :-) .
